src/include/type/expression/abstract_expression.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "type/expression/expression_defs.h"
+#include "type/type_id.h"
+
+namespace terrier {
+namespace type {
+namespace expression {
+
+/**
+ * An abstract parser expression. Dumb and immutable.
+ */
+class AbstractExpression {
+ protected:
+  /**
+   * Instantiates a new abstract expression. Because these are logical expressions, everything should be known
+   * at the time of instantiation, i.e. the resulting object is immutable.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param children the list of children for this node
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type,
+                     const std::vector<AbstractExpression *> children)
+      : expression_type_(expression_type), return_value_type_(return_value_type) {
+    for (auto &child : children) {
+      children_.emplace_back(std::unique_ptr<AbstractExpression>(child));
+      if (child->is_nullable_) {
+        is_nullable_ = true;
```

Comment:
What does is_nullable_ mean? I'm not clear, even after looking at Peloton and doing some reading.
1. Add comments to the declaration in abstract_expression.h to spell out what this is.
2. In Peloton, is_nullable is method that walks the expression tree. Here it is a flag that just looks at the first level child. I suspect that what is needed is the Peloton implementation.
3. Ditto for has_parameter_ and has_subquery_